cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -1276,6 +1496,65 @@ std::shared_ptr<ScalarFunction> MakeUnaryArithmeticFunctionNotNull(
   return func;
 }
 
+// Generate a kernel given an arithmetic rounding functor
+template <template <RoundMode> class Op>
+ArrayKernelExec GenerateArithmeticRound(RoundMode rmode, detail::GetTypeId ty) {
+  switch (rmode) {
+    case RoundMode::DOWN:
+      return GenerateArithmeticFloatingPoint<ScalarUnaryNotNull, Op<RoundMode::DOWN>>(ty);
+    case RoundMode::UP:
+      return GenerateArithmeticFloatingPoint<ScalarUnaryNotNull, Op<RoundMode::UP>>(ty);
+    case RoundMode::TOWARDS_ZERO:
+      return GenerateArithmeticFloatingPoint<ScalarUnaryNotNull,
+                                             Op<RoundMode::TOWARDS_ZERO>>(ty);
+    case RoundMode::TOWARDS_INFINITY:
+      return GenerateArithmeticFloatingPoint<ScalarUnaryNotNull,
+                                             Op<RoundMode::TOWARDS_INFINITY>>(ty);
+    case RoundMode::HALF_DOWN:
+      return GenerateArithmeticFloatingPoint<ScalarUnaryNotNull,
+                                             Op<RoundMode::HALF_DOWN>>(ty);
+    case RoundMode::HALF_UP:
+      return GenerateArithmeticFloatingPoint<ScalarUnaryNotNull, Op<RoundMode::HALF_UP>>(
+          ty);
+    case RoundMode::HALF_TOWARDS_ZERO:
+      return GenerateArithmeticFloatingPoint<ScalarUnaryNotNull,
+                                             Op<RoundMode::HALF_TOWARDS_ZERO>>(ty);
+    case RoundMode::HALF_TOWARDS_INFINITY:
+      return GenerateArithmeticFloatingPoint<ScalarUnaryNotNull,
+                                             Op<RoundMode::HALF_TOWARDS_INFINITY>>(ty);
+    case RoundMode::HALF_TO_EVEN:
+      return GenerateArithmeticFloatingPoint<ScalarUnaryNotNull,
+                                             Op<RoundMode::HALF_TO_EVEN>>(ty);
+    case RoundMode::HALF_TO_ODD:
+      return GenerateArithmeticFloatingPoint<ScalarUnaryNotNull,
+                                             Op<RoundMode::HALF_TO_ODD>>(ty);
+    default:
+      DCHECK(false);
+      return ExecFail;
+  }
+}
+
+// Like MakeUnaryArithmeticFunction, but for unary rounding functions that control
+// kernel dispatch based on RoundMode, only on non-null output.
+template <template <RoundMode> class Op, typename OptionsType>
+std::shared_ptr<ScalarFunction> MakeUnaryRoundFunction(std::string name,
+                                                       const FunctionDoc* doc) {
+  using State = RoundOptionsWrapper<OptionsType>;
+
+  static const OptionsType kDefaultOptions = OptionsType::Defaults();
+  auto func = std::make_shared<ArithmeticFloatingPointFunction>(name, Arity::Unary(), doc,
+                                                                &kDefaultOptions);
+  for (const auto& ty : FloatingPointTypes()) {
+    auto exec = [&](KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+      auto options = State::Get(ctx);
+      auto exec_ = GenerateArithmeticRound<Op>(options.round_mode, ty);
```

Comment:
Just for the record, it feels a bit weird to call `GenerateArithmeticRound` at kernel execution time. That said, a quick benchmarking in Python shows there doesn't seem to be any large overhead:
```python
>>> import pyarrow as pa, pyarrow.compute as pc
>>> floor = pc.get_function("floor")
>>> round = pc.get_function("round")
>>> arr = pa.array([None], type=pa.float64())
>>> %timeit floor.call([arr])
2.57 µs ± 10.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit floor.call([arr])
2.58 µs ± 11.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit round.call([arr])
2.65 µs ± 10.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit round.call([arr])
2.53 µs ± 12.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
I agree. The purpose of doing this here is to generate dispatchers once prior to kernel invocation. Previously, I tried 2 solutions:
* Have a vector of precompute `GenerateArithmeticFloatingPoint` and use the `options.round_mode` to index this vector. But this required vector to be ordered identically to the round modes in `enum RoundMode`.
* Similar to above but using an `unordered_map` indexed by `options.round_mode`. This requires adding hash support for `RoundMode` data type. This approach does not imposes a full constraint on the ordering of `RoundMode`.

Which one do you think is best?

Reply:
Actually, it is doing the same thing as the other kernels. Other kernels pass `exec` (an `ArrayKernelExec`) to the `AddKernel` method without invoking it. `exec` is invoked during kernel dispatching because it requires `KernelContext, ExecBatch, and Datum` parameters.